### PR TITLE
Orderbook/RPCServer: Fix GetOrderbook/Retrieve race condition (refactored and sped up)

### PR DESCRIPTION
**Pull request by vazha**
# PR Description

Please include a summary of the change, feature or issue which this pull request addresses. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

Because of function GetSpecificOrderbook return a pointer of orderbook.Base - there was a race condition in method GetOrderbook (package engine), and I got the panic quite often here.

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

**Review by shazbert — CHANGES_REQUESTED** [review]
Awesome pickup!

I think it might be better to change this: https://github.com/thrasher-corp/gocryptotrader/blob/master/exchanges/orderbook/orderbook.go#L162

So there's no chance of this happening to other subsystems. 

Even that copy of the ob.Bids and ob.Asks there is a potential race when the underlying pointer is being updated at the same time. So doing this update under a read lock should still ensure concurrency.

**Review by xtda — CHANGES_REQUESTED** [review]
Only thing on what of what was already suggested is to please remove the go_build_github_com_thrasher_corp_gocryptotrader binary as well

**Comment by vazha** [comment]
> I think it might be better to change this: https://github.com/thrasher-corp/gocryptotrader/blob/master/exchanges/orderbook/orderbook.go#L162
> So there's no chance of this happening to other subsystems.

Yes, good idea to make changes here.

> Even that copy of the ob.Bids and ob.Asks there is a potential race when the underlying pointer is being updated at the same time. So doing this update under a read lock should still ensure concurrency.

Dear shazbert, IMHO copy (localCopyOfAsks and localCopyOfBids) make a new slice which has it's own memory allocation and used only in read context, so it's impossible to get a race

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for making those changes! I was only meaning when we make a copy of the original slice which could be updating via websocket quite frequently there might be an issue there. But yeah I do see your point. Just one small nit. Then I can approve, fine work!

File: exchanges/orderbook/orderbook.go [hunk 1]
```diff
@@ -140,26 +140,41 @@ func (s *Service) Retrieve(exchange string, p currency.Pair, a asset.Item) (*Bas
...
 	}
 
-	if s.Books[exchange][p.Base.Item][p.Quote.Item] == nil {
+	if _, ok := s.Books[exchange][p.Base.Item][p.Quote.Item]; !ok {
 		return nil, fmt.Errorf("no orderbooks associated with quote currency %s",
 			p.Quote)
 	}
 
-	if s.Books[exchange][p.Base.Item][p.Quote.Item][a] == nil {
+	if _, ok := s.Books[exchange][p.Base.Item][p.Quote.Item][a]; !ok {
```
[the remark this answers is not in the document]
Comment (vazha) [reply exchanges/orderbook/orderbook.go]: Agree with you, that will be better

**Review by shazbert — APPROVED** [review]
Just noticed an issue with LastUpdated not being updated after the first insertion but that is Out of scope for this PR.
Tested Approved.

**Review by MadCozBadd — APPROVED** [review]
TACK

File: engine/rpcserver.go [hunk 2]
```diff
@@ -401,20 +401,26 @@ func (s *RPCServer) GetOrderbook(_ context.Context, r *gctrpc.GetOrderbookReques
...
 
 	var bids []*gctrpc.OrderbookItem
-	for x := range ob.Bids {
-		bids = append(bids, &gctrpc.OrderbookItem{
-			Amount: ob.Bids[x].Amount,
-			Price:  ob.Bids[x].Price,
-		})
-	}
-
 	var asks []*gctrpc.OrderbookItem
```
Comment (xtda) [inline engine/rpcserver.go:404]: You can make() these with the len() of ob.Bids / ob.Asks if you want since we append them all anyway and will save having to increase the size of the slice as part of the append()

You can also do assignment  instead of append in the loops below if you really want but the performance gains are minimal/non-existent that its whatever you consider more readable

**Review by xtda — APPROVED** [review]
tested approved with an optional small comment

**Review by gloriousCode — APPROVED** [review]
Thanks for finding another sweet bug and fixing it @vazha! Tested and all good

**Review by thrasher- — CHANGES_REQUESTED** [review]
When testing GetOrderbookStream GCTCLI command, it appears that the orderbook is not reflecting changes correctly on this PR compared to master.

image

You can see that in this picture, the orderbook asks pricing are < then the bids. For testing, I used BTC-USDT,ETH-USDT,LTC-USDT,ADA-USDT,XRP-USDT as enabled pairs with websocket enabled.

**Comment by vazha** [comment]
> When testing GetOrderbookStream GCTCLI command, it appears that the orderbook is not reflecting changes correctly on this PR compared to master.
> 
> image
> 
> You can see that in this picture, the orderbook asks pricing are < then the bids. For testing, I used BTC-USDT,ETH-USDT,LTC-USDT,ADA-USDT,XRP-USDT as enabled pairs with websocket enabled.

Yes, yes, I found this issue earlier (before thit PR) and still work on this for next PR. I'll give you the details later today. Thank you!

**Comment by thrasher-** [comment]
> > When testing GetOrderbookStream GCTCLI command, it appears that the orderbook is not reflecting changes correctly on this PR compared to master.
> > You can see that in this picture, the orderbook asks pricing are < then the bids. For testing, I used BTC-USDT,ETH-USDT,LTC-USDT,ADA-USDT,XRP-USDT as enabled pairs with websocket enabled.
> 
> Yes, yes, I found this issue earlier (before thit PR) and still work on this for next PR. I'll give you the details later today. Thank you!

Thanks @vazha !

**Comment by vazha** [comment]
> > > When testing GetOrderbookStream GCTCLI command, it appears that the orderbook is not reflecting changes correctly on this PR compared to master.
> > > You can see that in this picture, the orderbook asks pricing are < then the bids. For testing, I used BTC-USDT,ETH-USDT,LTC-USDT,ADA-USDT,XRP-USDT as enabled pairs with websocket enabled.
> > 
> > 
> > Yes, yes, I found this issue earlier (before thit PR) and still work on this for next PR. I'll give you the details later today. Thank you!
> 
> Thanks @vazha !



> > > When testing GetOrderbookStream GCTCLI command, it appears that the orderbook is not reflecting changes correctly on this PR compared to master.
> > > You can see that in this picture, the orderbook asks pricing are < then the bids. For testing, I used BTC-USDT,ETH-USDT,LTC-USDT,ADA-USDT,XRP-USDT as enabled pairs with websocket enabled.
> > 
> > 
> > Yes, yes, I found this issue earlier (before thit PR) and still work on this for next PR. I'll give you the details later today. Thank you!
> 
> Thanks @vazha !

Hi guys, researching the problem took me longer than I thought. Here is the conclusion I came to (here we are talking about the Binance implementation, I have not yet check another exchanges): 
at the time the program is start (binance_websocket.go, WsConnect method), the b.SeedLocalCache (enabledPairs [i]) is executed for each pair, but the problem is that during this time - synchronization is not performed, 
and a few seconds later, when reached go b.wsReadData() - LocalBuffer may no longer be actual and later a situation arises when asks < bids.

As a solution - I am thinking of doing something like:
exchanges/binance/binance_websocket.go:

	enabledPairs, err := b.GetEnabledPairs(asset.Spot)
	if err != nil {
		return err
	}

	go b.wsReadData()

	subs, err := b.GenerateSubscriptions()
	if err != nil {
		return err
	}

	err = b.Websocket.SubscribeToChannels(subs)
	if err != nil {
		return err
	}

	for i := range enabledPairs {
		fmt.Println("SeedLocalCache_1", enabledPairs[i])
		err = b.SeedLocalCache(enabledPairs[i])
		if err != nil {
			return err
		}
	}

	return nil
}

that the synchronization starts before the initial snapshot is loaded.

To summarize, the problem is not in this PR, and I think that we can close it and move on.

**Review by gloriousCode — APPROVED** [review]
Binance is quite tricky with the timings for websocket. It would be nice if they were like other exchanges and sent a snapshot via websocket...... But anyway, it looks like we'll have to come up with a more detailed solution. Thank you for investigating that!

I retested this against Kraken and saw that it matched correctly and that I also did not experience a data race panic constantly retrieving the orderbook. Thank you again @vazha

**Review by shazbert — APPROVED** [review]
tACK new changes.

**Review by MadCozBadd — APPROVED** [review]
LGTM!!! TACK

**Comment by thrasher-** [comment]
> > > > When testing GetOrderbookStream GCTCLI command, it appears that the orderbook is not reflecting changes correctly on this PR compared to master.
> > > > You can see that in this picture, the orderbook asks pricing are < then the bids. For testing, I used BTC-USDT,ETH-USDT,LTC-USDT,ADA-USDT,XRP-USDT as enabled pairs with websocket enabled.
> > > 
> > > 
> > > Yes, yes, I found this issue earlier (before thit PR) and still work on this for next PR. I'll give you the details later today. Thank you!
> > 
> > 
> > Thanks @vazha !
> 
> > > > When testing GetOrderbookStream GCTCLI command, it appears that the orderbook is not reflecting changes correctly on this PR compared to master.
> > > > You can see that in this picture, the orderbook asks pricing are < then the bids. For testing, I used BTC-USDT,ETH-USDT,LTC-USDT,ADA-USDT,XRP-USDT as enabled pairs with websocket enabled.
> > > 
> > > 
> > > Yes, yes, I found this issue earlier (before thit PR) and still work on this for next PR. I'll give you the details later today. Thank you!
> > 
> > 
> > Thanks @vazha !
> 
> Hi guys, researching the problem took me longer than I thought. Here is the conclusion I came to (here we are talking about the Binance implementation, I have not yet check another exchanges):
> at the time the program is start (binance_websocket.go, WsConnect method), the b.SeedLocalCache (enabledPairs [i]) is executed for each pair, but the problem is that during this time - synchronization is not performed,
> and a few seconds later, when reached go b.wsReadData() - LocalBuffer may no longer be actual and later a situation arises when asks < bids.
> 
> As a solution - I am thinking of doing something like:
> exchanges/binance/binance_websocket.go:
> 
> ```
> enabledPairs, err := b.GetEnabledPairs(asset.Spot)
> if err != nil {
> 	return err
> }
> 
> go b.wsReadData()
> 
> subs, err := b.GenerateSubscriptions()
> if err != nil {
> 	return err
> }
> 
> err = b.Websocket.SubscribeToChannels(subs)
> if err != nil {
> 	return err
> }
> 
> for i := range enabledPairs {
> 	fmt.Println("SeedLocalCache_1", enabledPairs[i])
> 	err = b.SeedLocalCache(enabledPairs[i])
> 	if err != nil {
> 		return err
> 	}
> }
> 
> return nil
> ```
> 
> }
> 
> that the synchronization starts before the initial snapshot is loaded.
> 
> To summarize, the problem is not in this PR, and I think that we can close it and move on.

Agreed, attempted Kraken orderbook streaming and it was fine so the Binance issue will have to be worked on in a different PR. Thanks @vazha !

**Review by thrasher- — APPROVED** [review]
